### DISCUSSION TITLE
Need some help to understand the behavior of Index trait on StringRecord

### DISCUSSION THREAD

**Discussion by meretciel** [BurntSushi/rust-csv#334] [Q&A] [answered]
rust version: rustc 1.71.0 (8ede3aae2 2023-07-12)
rust-csv version: 1.2.2

I don't know if I read the error message correctly, it seems that the index operator of `StringBuilder` class returns `str` instead of `&str`. Moreover, the implementation of the `index` function just calls the `get` method and the `get` returns `&str`. ([see code here](https://github.com/BurntSushi/rust-csv/blob/master/src/string_record.rs#L660)).

Sample Code:

```
    let mut reader = ReaderBuilder::new().from_path("your-csv-file.csv").unwrap();
    let header = reader.headers().unwrap();
    let t = header[0];  // This does not build, see attached error message below. If this line is 
                                  // commented out, then the program can build.
    let t2 = header.get(0);
```

It give the following error:

```
164 |     let t = header[0];
    |         ^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `str`
    = note: all local variables must have a statically known size
    = help: unsized locals are gated as an unstable feature
help: consider borrowing here
    |
164 |     let t = &header[0];
    |    
```

I'm wondering if there is any explanation of this behavior (It's surprising that we can even return a `str` in a function). Appreciate any helps in advance. Thanks!

**Comment by BurntSushi** [comment, marked as answer]
The error is telling you what you should do: use `&header[0]` instead of `header[0]`. The key here is that `header[0]` is actually syntactic sugar for `*header.index(0)`. Since the `Index` impl for `StringRecord` returns a `&str`, it follows that the type of `*header.index(0)` is `str`. Since `str` is a dynamically sized type, you get a compiler error here because the compiler doesn't know how much stack space to allocate for `t`.

This quirk tends to only apply to `Index` impls that return fat pointers, such as `&str`. In cases where the `Index` impl returns a regular pointer, say, `&u8`, then `slice[0]` would just return a `u8` which is typically what you want.

The docs for [`std::ops::Index`](https://doc.rust-lang.org/std/ops/trait.Index.html) are a little sparse, but they might help a little.

**Reply by meretciel** [reply]
Thank you so much for the explanation! I didn't realize `container[index] is actually syntactic sugar for *container.index(index)`.
